controllers/storagecluster/external_ocs.go

```diff
@@ -0,0 +1,221 @@
+package storagecluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v1"
+	externalClient "github.com/red-hat-storage/ocs-operator/services/provider/client"
+)
+
+var (
+	// externalOCSResources will hold the ExternalResources for storageclusters
+	// ExternalResources can be accessible using the UID of an storagecluster
+	externalOCSResources = map[types.UID][]ExternalResource{}
+)
+
+// isExternalOCSProvider returns true if it is ocs to ocs ExternalStorage consumer cluster
+func isExternalOCSProvider(instance *ocsv1.StorageCluster) bool {
+	return instance.Spec.ExternalStorage.Enable && instance.Spec.ExternalStorage.StorageProviderKind == ocsv1.KindOCS
+}
+
+// newExternalClusterClient returns the *grpc.ClientConn and *providerClient.ConsumerClient
+func newExternalClusterClient(instance *ocsv1.StorageCluster) (*grpc.ClientConn, *externalClient.OCSProviderClient, error) {
+
+	// TODO: implement logic for separating address and token
+
+	grpcConnection, err := externalClient.NewGRPCConnection(
+		instance.Spec.ExternalStorage.ConnectionString, []grpc.DialOption{})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	consumerClient := externalClient.NewProviderClient(grpcConnection, time.Second*10)
+	return grpcConnection, consumerClient, nil
+}
+
+// onboardConsumer makes an API call to the external storage provider cluster for onboarding
+func onboardConsumer(r *StorageClusterReconciler, instance *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	clusterVersion := &configv1.ClusterVersion{}
+	err := r.Client.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
```

Comment:
We should not be creating new context objects for each client call. there should be a single context object for an entire reconcile run that is created on the main reconcile function and provided here from the outside.

 Also please use context.Background() to create that context. and not context.TODO().

Reply:
I can use context.Background() but won't be able to pass the one coming from reconcile.

Reply:
added context.Background()

Reply:
I think the main ask was to have a single context for all of the calls, it is the important part of the ask.
Just accept it as a parameter

Reply:
In that case, I have to change the interface again to pass the context and use the context as an argument, I think we can take this one later.